File: 0726-number-of-atoms/0726-number-of-atoms.py
class Solution:
    def countOfAtoms(self, formula: str) -> str:
        tokens = re.findall(r'(\(|\)|[A-Z][a-z]*)(\d*)', formula)
        
        stack = [defaultdict(int)]
        
        for element, count in tokens:
            if element == '(':
                stack.append(defaultdict(int))
            elif element == ')':
                last_counts = stack.pop()
                multiplier = int(count) if count else 1
                for key, value in last_counts.items():
                    stack[-1][key] += value * multiplier
            else:
                count = int(count) if count else 1
                stack[-1][element] += count
        
        final_counts = sorted(stack.pop().items())
        return ''.join([f"{element}{count if count > 1 else ''}" for element, count in final_counts])        

File: 0726-number-of-atoms/0726-number-of-atoms.rb
# @param {String} formula
# @return {String}
def count_of_atoms(formula)
  tokens = formula.scan(/(\W|[A-Z][a-z]*)(\d*)/)
  
  stack = tokens.reduce([Hash.new(0)]) do |stack, (element, count)|
    if element == '('
      stack << Hash.new(0)
    elsif element == ')'
      last_counts = stack.pop
      multiplier = count.to_i > 0 ? count.to_i : 1

      last_counts.each do |key, value|
        stack.last[key] += value * multiplier
      end
      
    else
      count = count.to_i > 0 ? count.to_i : 1
      stack.last[element] += count
    end

    stack
  end

  final_counts = stack.pop.sort
  final_counts.map { |element, count| count > 1 ? "#{element}#{count}" : element }.join
end
